### ISSUE TITLE
chromedriver version not match with browser

### ISSUE THREAD

**Issue by LeeYellow** [heroku/heroku-buildpack-chromedriver#47]
My app crashed today with the following error:
selenium.common.exceptions.SessionNotCreatedException: Message: session not created: This version of ChromeDriver only supports Chrome version 114
Current browser version is 116.0.5845.96 with binary path /app/.apt/opt/google/chrome/chrome

I set the env variable CHROMEDRIVER_VERSION to (and re-deploy the app)
114.0.5735.16
or
116.0.5845.96

When I set the env variable CHROMEDRIVER_VERSION to 114.0.5735.16
here popping up the message:
**chromedriver app detected**
**remote: -----> Downloading chromedriver v114.0.5735.16...**
and then still the same error message pops when I call chrome driver.
selenium.common.exceptions.SessionNotCreatedException: Message: session not created: This version of ChromeDriver only supports Chrome version 114
Current browser version is 116.0.5845.96 with binary path /app/.apt/opt/google/chrome/chrome

When I set the env variable CHROMEDRIVER_VERSION to 116.0.5845.96 
heroku reject me to deploy again
here is the error message:

remote:   End-of-central-directory signature not found.  Either this file is not
remote:   a zipfile, or it constitutes one disk of a multi-part archive.  In the
remote:   latter case the central directory and zipfile comment will be found on
remote:   the last disk(s) of this archive.
remote: unzip:  cannot find zipfile directory in one of /tmp/chromedriver.zip or
remote:         /tmp/chromedriver.zip.zip, and cannot find /tmp/chromedriver.zip.ZIP, period.
remote:  !     Push rejected, failed to compile chromedriver app.
remote:
remote:  !     Push failed
remote: Verifying deploy...
remote:
remote: !       Push rejected to ema-cloz.
remote:
To https://git.heroku.com/XXXXXXX

does anyone run into the same issue?
thanks

**Comment by ghost** [comment]
yes same issue, https://googlechromelabs.github.io/chrome-for-testing/ the chromedrivers after 114 changed to new domain
PR: https://github.com/heroku/heroku-buildpack-chromedriver/pull/45/files

**Comment by LeeYellow** [comment]
@lavarthan Thanks for the information.
any idea to solve this problem? I want to downgrade the browser version, but I could not find the way under heroku environment.

**Comment by ghost** [comment]
Until they approve the PR, you can use this build pack
https://github.com/lavarthan/heroku-build-pack-chrome-driver
I have added the fix

**Comment by chbrown1293** [comment]
Im trying to use this buildpack but getting this error. Any ideas?

`
2023-08-16 14:53:11 WARN Selenium [:selenium_manager] Error getting version of chromedriver 116. Retrying with chromedriver 115 (attempt 1/5) 
2023-08-16 14:53:11 WARN Selenium [:selenium_manager] Error getting version of chromedriver 115. Retrying with chromedriver 114 (attempt 2/5) 
#0 0x55e97da654e3 <unknown>: session not created: This version of ChromeDriver only supports Chrome version 114 (Selenium::WebDriver::Error::SessionNotCreatedError)
Current browser version is 116.0.5845.96 with binary path /app/.apt/usr/bin/google-chrome`

**Comment by ghost** [comment]
@chbrown1293  which build pack are you using?
this one working fine https://github.com/lavarthan/heroku-build-pack-chrome-driver

**Comment by chbrown1293** [comment]
Same. 

`heroku buildpacks:add https://github.com/lavarthan/heroku-build-pack-chrome-driver`

**Comment by ghost** [comment]
I'm using this without any problem, may be purge the cache and try

**Comment by chbrown1293** [comment]
Ive been trying that as well. Strange.

**Comment by chbrown1293** [comment]
Same thing when creating a new app with it, so I dont think its the cache.

**Comment by ghost** [comment]
@chbrown1293  are you using ChromeDriverManager inside the code to install the driver?

**Comment by chbrown1293** [comment]
I am not, just the heroku cli

**Comment by KulovacNedim** [comment]
The same here. 
Also, https://github.com/lavarthan/heroku-build-pack-chrome-driver is not working for me.

**Comment by fonmagnus** [comment]
I encounter the same issue

**Comment by fonmagnus** [comment]
I use your solution but got an error message like this

```
remote: -----> chromedriver app detected
remote: -----> Downloading chromedriver v114.0.5735.90...
remote: Archive:  /tmp/chromedriver.zip
remote:   End-of-central-directory signature not found.  Either this file is not
remote:   a zipfile, or it constitutes one disk of a multi-part archive.  In the
remote:   latter case the central directory and zipfile comment will be found on
remote:   the last disk(s) of this archive.
remote: unzip:  cannot find zipfile directory in one of /tmp/chromedriver.zip or
remote:         /tmp/chromedriver.zip.zip, and cannot find /tmp/chromedriver.zip.ZIP, period.
remote:  !     Push rejected, failed to compile chromedriver app.
remote: 
remote:  !     Push failed
remote: Verifying deploy...
```

**Comment by srehne** [comment]
I could deploy the 
https://github.com/lavarthan/heroku-build-pack-chrome-driver 
solution without errors ... 
But sadly I got an error when trying to use it ...
`Unable to find chromedriver. Please download the server from\nhttps://chromedriver.storage.googleapis.com/index.html and place it somewhere on your PATH.
More info at https://github.com/SeleniumHQ/selenium/wiki/ChromeDriver.",`

I also tried to add the webdriver gem in the production env (which I only used in development before).
But got the same problem

I know that I am 'shooting in all directions' ... but this is a big issue for us!! 

I know we might need to do a larger upgrade ... but it would be nice if it was possible to stay on an older CHROMEDRIVER_VERSION so small changes don't result in need to change the CHROMEDRIVER_VERSION.

**Comment by fonmagnus** [comment]
i agree. somehow the chromedriver keeps installing version 116 instead of 114 and this is a breaking changes to the selenium i run in the python application

**Comment by OlivierLej** [comment]
Hello, for the record, we're facing the same issue. 

We'd be interested in switching to this buildpack: https://github.com/lavarthan/heroku-build-pack-chrome-driver

Is there a list of steps required to make this work?

For example, our current CHROMEDRIVER_PATH env variable is set to /app/.chromedriver/bin/chromedriver

Does it need to change?

Appreciate your help

**Comment by ghost** [comment]
> Hello, for the record, we're facing the same issue.
> 
> We'd be interested in switching to this buildpack: https://github.com/lavarthan/heroku-build-pack-chrome-driver
> 
> Is there a list of steps required to make this work?
> 
> For example, our current CHROMEDRIVER_PATH env variable is set to /app/.chromedriver/bin/chromedriver
> 
> Does it need to change?
> 
> Appreciate your help

hope this is [the build pack](https://github.com/heroku/heroku-buildpack-google-chrome.git) you are using for the google chrome
no need to explicitly set the chromedriver path, just set the chrome binary location as `/app/.apt/usr/bin/google-chrome`

```
options = webdriver.ChromeOptions()
options.binary_location = "/app/.apt/usr/bin/google-chrome"
// rest of the options
service = ChromeService()
driver = webdriver.Chrome(service=service, options=options)
```

**Comment by srehne** [comment]
> I could deploy the https://github.com/lavarthan/heroku-build-pack-chrome-driver solution without errors ... But sadly I got an error when trying to use it ... `Unable to find chromedriver. Please download the server from\nhttps://chromedriver.storage.googleapis.com/index.html and place it somewhere on your PATH. More info at https://github.com/SeleniumHQ/selenium/wiki/ChromeDriver.",`
> 
> I also tried to add the webdriver gem in the production env (which I only used in development before). But got the same problem
> 
> I know that I am 'shooting in all directions' ... but this is a big issue for us!!
> 
> I know we might need to do a larger upgrade ... but it would be nice if it was possible to stay on an older CHROMEDRIVER_VERSION so small changes don't result in need to change the CHROMEDRIVER_VERSION.

Last nights update made everything work for me. 
I updated to the latest selenium-webdriver 4.11.0 and things are now working! 
Thanks to "the team" / johlym !

**Comment by nphenley** [comment]
I am currently using version 114.0.5735.90 of Chrome and ChromeDriver. I am able to set the version number for chrome driver with the default heroku buildpack https://github.com/heroku/heroku-buildpack-chromedriver.git 

When Chrome updated to 116, it broke my app because ChromeDriver didn't have a version 116 I could use. I therefore started using this fork https://github.com/awl19/heroku-buildpack-google-chrome.git that allows you to set the Chrome version.

This was working fine up until last night but now, whenever I try to deploy a new build I get this issue:
`-----> Downloading chromedriver v114.0.5735.90...
Archive:  /tmp/chromedriver.zip
  End-of-central-directory signature not found.  Either this file is not
  a zipfile, or it constitutes one disk of a multi-part archive.  In the
  latter case the central directory and zipfile comment will be found on
  the last disk(s) of this archive.
unzip:  cannot find zipfile directory in one of /tmp/chromedriver.zip or
        /tmp/chromedriver.zip.zip, and cannot find /tmp/chromedriver.zip.ZIP, period.
 !     Push rejected, failed to compile chromedriver app.
 !     Push failed`

According to above, it seems the chrome driver download URL that the buildpack uses has changed. For some reason, my version number can't be found with the new link https://edgedl.me.gvt1.com/edgedl/chrome/chrome-for-testing/114.0.5735.90/linux64/chromedriver-linux64.zip

This seems to be the issue as far as I can tell. Does anyone have any suggestions?

**Comment by ghost** [comment]
@nphenley  this url of the chromedriver only work after 114 version, now they have updated the domain url so only after the chrome version 114 will work.(see the latest [PR](https://github.com/heroku/heroku-buildpack-chromedriver/pull/51) merged in the repo)

**Comment by nphenley** [comment]
@lavarthan oh OK thanks. So I guess now I can just use the default heroku buildpacks, both for chrome and chrome driver and just let them use the most recent 116 version?

**Comment by ghost** [comment]
@nphenley Yes

**Comment by edmorley** [comment]
Support for newer chromedriver versions was fixed by #51 + #52.

Note: Older versions may no longer be available from the new download site. If you see errors, try unsetting `CHROMEDRIVER_VERSION` to use the default latest version.
